### DISCUSSION TITLE
Niimbot Label Printer Integration?

### DISCUSSION THREAD

**Discussion by cobaltix-max** [Donkie/Spoolman#320] [Ideas]
Hey everyone, apologies if this isn't how to do this (though it certainly seems like the place). I recently got spoolman up and running, and I'm extremely grateful for this project!

I bought a very inexpensive thermal label printer to label all my spools, and it's geared towards crappy mobile app usage, but it also has a USB interface that presents as a serial device. Zero docs that I can find, and they would likely be in chinese, but I stumbled upon a python project that allows usage of the device directly without BS apps. I am **not** a developer, but it occurred to me that this would be such a perfect inexpensive solution for many spoolman users to label their spools, and it would be very slick to be able to effectively churn out labels directly from the web interface., 

Here's the printer: https://www.niimbot.net/enweb/goods_detail.html?goods_id=68
Here's the python project: https://github.com/AndBondStyle/niimprint

Of course, I will likely try my hand at simply putting together something that renders labels from Spoolman CSV output and pushes it to the printer, but maybe I'll actually finish putting the skirts on my voron v2 before I pretend to be a developer.. 🥇

**Comment by laundry-drivel** [comment]
Hi is there any update on how you did this? I also have a Niimbot printer and struggling to work out how I can print with this

**Comment by bliepp** [comment]
There's also a [Home Assistant integration](https://github.com/eigger/hass-niimbot) by the community, which basically wraps [niimprint](https://github.com/AndBondStyle/niimprint). I didn't test it due to the lack of a Niimbot device, but as far as I understand it, it's as easy as creating an automation in Home Assistant that's responding to a webhook.

Unfortunately, spoolman doesn't have support for webhooks (yet?), but that shouldn't be hard at all to implement (at least from glancing over the source code). Also, using webhooks instead of directly implementing a printer has many advantages:
1. the printer can be basically anywhere (even on a remote location like a storage room) and isn't physically bound to the host machine
2. you could swap this for any other webhook receiver you want or need (like just some other fancy home automation, or printing with any other printer - basically any CLI command with the [webhook](https://github.com/adnanh/webhook/) project)
3. the printer can be used for more than just printing from spoolman, it could also be used with grocy or just the manually invoked

So for now, I would say webhooks are the way to go and it's up to the community/individual to come up with an appropriate solution. I think spoolman shouldn't be (directly) bothered with such implementation details. Other inventory systems like [grocy](https://grocy.info/de) use that approach and I think it would be appropriate adapt that.
